### PR TITLE
libsed: Make opal level0 discovery struct sed device context

### PR DESCRIPTION
**Pull request by RevanthRajashekar**
Make level0 discovery struct sed device context rather than an
individual struct avoiding any over-written conditions.

Move the sed_device struct from sed_util library to libsed library.

Signed-off-by: Revanth Rajashekar <revanth.rajashekar@intel.com>

### CODE REVIEW

File: src/lib/include/libsed.h [hunk 1]
```diff
@@ -105,6 +103,12 @@ struct sed_opal_level0_discovery {
 	struct sed_opalv200_supported_feat sed_opalv200;
 };
 
+struct sed_device {
```
Comment (AndrzejJakowski) [inline src/lib/include/libsed.h]: I would like to keep that structure private, w/o defining its fields to callers.

File: src/lib/include/libsed.h [hunk 2]
```diff
@@ -159,7 +163,7 @@ int sed_init(struct sed_device **dev, const char *dev_path);
 /**
  *
  */
-int sed_level0_discovery(struct sed_opal_level0_discovery *discv);
```
Comment (AndrzejJakowski) [inline src/lib/include/libsed.h:162]: I like original prototype of this function which takes discovery data structure and populates it w/ data.

File: src/lib/include/libsed.h [hunk 3]
```diff
@@ -159,7 +163,7 @@ int sed_init(struct sed_device **dev, const char *dev_path);
...
+int sed_level0_discovery(void);
```
Comment (AndrzejJakowski) [inline src/lib/include/libsed.h]: this seems to have no effect

File: src/lib/include/libsed.h (hunk 1 above)
Comment (RevanthRajashekar) [reply src/lib/include/libsed.h]: Done

File: src/lib/include/libsed.h (hunk 2 above)
Comment (RevanthRajashekar) [reply src/lib/include/libsed.h:162]: Done

File: src/lib/nvme_pt_ioctl.c [hunk 4]
```diff
@@ -414,9 +413,18 @@ int opal_init_pt(struct sed_device *dev, const char *device_path)
...
+				struct sed_opal_level0_discovery *discv)
 {
-	memcpy(discvry, (struct sed_opal_level0_discovery *)discv, sizeof(*discvry));
+	int ret;
+
+	ret = opal_level0_disc_pt(dev);
+	if (ret)
+		return ret;
+
+	memcpy(discv, &dev->discv, sizeof(*discv));
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c:428]: before doing memcpy you could check if if input is not null

File: src/lib/nvme_pt_ioctl.c [hunk 5]
```diff
@@ -200,42 +198,57 @@ static int get_opal_auth_uid(enum SED_AUTHORITY auth)
 	return sed2opal_map[auth];
 }
 
-static void check_tper_feat(void *feat)
+static void check_tper_feat(struct sed_opal_level0_discovery *discv,
```
Comment (AndrzejJakowski) [inline src/lib/nvme_pt_ioctl.c]: function naming is a little bit confusing: it doesn't check anything it just perform memcpy

File: src/sedcli_main.c [hunk 6]
```diff
@@ -761,37 +751,36 @@ static int handle_setpw(void)
...
+	ret = sed_level0_discovery(dev, &discv);
 	if (ret) {
-		if (ret == -EOPNOTSUPP)
-			sedcli_printf(LOG_ERR, "Command NOT supported for this "
-					"interface.\n");
-		else
-			sedcli_printf(LOG_ERR, "Error in level0 discovery\n");
-		goto free_mem;
+		if (ret == -EOPNOTSUPP) {
+			sedcli_printf(LOG_ERR, "Level0 discovery not supported "
```
Comment (AndrzejJakowski) [inline src/sedcli_main.c]: sounds like this is should be logged as LOG_WARNING vs LOG_ERR, because you allow operations

**Review by AndrzejJakowski — COMMENTED** [review]
Besides this few minor comments this looks good to me

File: src/lib/nvme_pt_ioctl.c (hunk 5 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c]: yup, will put it in another commit

File: src/lib/nvme_pt_ioctl.c (hunk 4 above)
Comment (RevanthRajashekar) [reply src/lib/nvme_pt_ioctl.c:428]: sorry for missing out this check condition

File: src/sedcli_main.c (hunk 6 above)
Comment (RevanthRajashekar) [reply src/sedcli_main.c]: will do it

**Comment by RevanthRajashekar** [comment]
Changes incorporated based on comments. Please review

**Pull request merged**